examples/step-78/step-78.cc

```diff
@@ -0,0 +1,869 @@
+  template <int dim>
+  void BlackScholes<dim>::o_results()
+  {
+    data_out_stack.new_parameter_value(time, time_step);
+    data_out_stack.attach_dof_handler(dof_handler);
+    data_out_stack.add_data_vector(solution, solution_names);
+    data_out_stack.build_patches(2);
+    data_out_stack.finish_parameter_value();
+  }
+
+  /*
+  This is the function that refines the spatial mesh. This is exactly what was
+  used in Step-26, so there is nothing new here.
+  */
+  template <int dim>
+  void BlackScholes<dim>::refine_mesh(const unsigned int min_grid_level,
+                                      const unsigned int max_grid_level)
+  {
+    Vector<float> estimated_error_per_cell(triangulation.n_active_cells());
+    KellyErrorEstimator<dim>::estimate(
+      dof_handler,
+      QGauss<dim - 1>(fe.degree + 1),
+      std::map<types::boundary_id, const Function<dim> *>(),
+      solution,
+      estimated_error_per_cell);
+    GridRefinement::refine_and_coarsen_fixed_fraction(triangulation,
+                                                      estimated_error_per_cell,
+                                                      0.6,
+                                                      0.4);
+    if (triangulation.n_levels() > max_grid_level)
+      for (const auto &cell :
+           triangulation.active_cell_iterators_on_level(max_grid_level))
+        cell->clear_refine_flag();
+    for (const auto &cell :
+         triangulation.active_cell_iterators_on_level(min_grid_level))
+      cell->clear_coarsen_flag();
+    SolutionTransfer<dim> solution_trans(dof_handler);
+    Vector<double>        previous_solution;
+    previous_solution = solution;
+    triangulation.prepare_coarsening_and_refinement();
+    solution_trans.prepare_for_coarsening_and_refinement(previous_solution);
+    triangulation.execute_coarsening_and_refinement();
+    setup_system();
+    solution_trans.interpolate(previous_solution, solution);
+    constraints.distribute(solution);
+  }
+
+  /*
+  This is somewhat unnecessary to have a function for the global refinement that
+  we do. The reason for the function, is to allow for the possibility of an
+  adaptive refinement later.
+  */
+
+  template <int dim>
+  void BlackScholes<dim>::refine_grid()
+  {
+    triangulation.refine_global(1);
+  }
+
+  /*
+  Now we get to the main driver of the program. This is where we do all the work
+  of looping through the timesteps and calculating the solution vector each
+  time. Here at the top, we set the initial refinement value and then create a
+  mesh. Then we refine this mesh once. Next, we set up the data_out_stack object
+  to store our solution. Finally we create a label for where we should start
+  when rerunning the first timestep and interpolate the initial condition onto
+  the mesh.
+  */
+  template <int dim>
+  void BlackScholes<dim>::run()
+  {
+    const unsigned int initial_global_refinement = 0; // 5 8
+
+    GridGenerator::hyper_cube(triangulation, 0.0, s_max);
```

Comment:
Why not `GridGenerator::hyper_cube(triangulation, 0.0, s_max, true);`?

Reply:
A general question. Would the code also work for 2 and 3D?

Reply:
I think I didn't use true here, because in 1D it always colorizes. I changed it just now though so that it would be this way if we elevated it. For now, it doesn't work in higher dimensions because I'm not implementing the general formula that counts for the correlation between assets. I would like to do this soon, but I don't think it'll be ready in the next few days.